### PR TITLE
Replace std::mutex with QD lock in write buffer

### PR DESCRIPTION
**Pull request by lundgren87**
This PR proposes to replace `std::mutex` with a **QD lock** in the ArgoDSM write buffer. To achieve this, the [QD library](https://github.com/davidklaftenegger/qd_library) is added as a submodule of ArgoDSM and is automatically fetched by CMake during the build procedure.

The goal of using a QD lock instead of `std::mutex` is to allow delegation to the current lock holder of the `write_buffer::add`, `write_buffer::erase` and `write_buffer::flush` operations. The benefit of delegating these operations will not be apparent until the removal of both the global `cachemutex` and `ibsem`, as these currently serialize write buffer operations regardless. With future removal of these restrictions, addition to or deletion from the write buffer can be done asynchronously without awaiting completion should the write buffer lock currently be held by someone else.

### CODE REVIEW

**Comment by lundgren87** [comment]
I've submitted a PR to the original QD library repo to build shared libraries. Once David has had time to review/merge it I will update this to the correct repo/branch and test on Jenkins.

**Review by kostis — APPROVED** [review]
Looks OK.

**Comment by lundgren87** [comment]
Changed repository to [davidklaftenegger/qd_library](https://github.com/davidklaftenegger/qd_library) tracking the `master` branch with the latest push. Should probably run on Jenkins to check this doesn't break anything before merging, but Gogs (hosting git repos needed by Jenkins) currently appears to be down.

**Comment by lundgren87** [comment]
This passed multiple runs on Jenkins and provides a small _overall_ performance boost for the Jenkins benchmarks.

File: src/backend/mpi/write_buffer.hpp
Comment (davidklaftenegger) [inline src/backend/mpi/write_buffer.hpp:445]: is there, or should there be a statistic for how much we burn CPU time here? High amounts of time spent in this line could indicate we are preventing other threads from doing useful work.

File: src/backend/mpi/write_buffer.hpp
Comment (davidklaftenegger) [inline src/backend/mpi/write_buffer.hpp:285]: code duplication with end of function. Please find a way to not repeat yourself.

**Review by davidklaftenegger — COMMENTED** [review]
Overall looks very good.

The helper functions are currently needed due to a limitation/bug in the QD library's API, so we can remove them once that is fixed, but for the difficult parts of this patch look good to me.

Some potential improvements are noted inline.

File: src/backend/mpi/write_buffer.hpp
Comment (lundgren87) [reply src/backend/mpi/write_buffer.hpp:174]: None other than simplicity as it is already used across the backend. If `MPI_Wtime()` is a bad choice I suggest we create an issue to move to `std::chrono` everywhere instead.

File: src/backend/mpi/write_buffer.hpp
Comment (davidklaftenegger) [reply src/backend/mpi/write_buffer.hpp]: What about `_page_count` and `_partial_flush_count`?

Comment (lundgren87) [reply src/backend/mpi/write_buffer.hpp]: Updated for these as well.

File: src/backend/mpi/write_buffer.hpp
Comment (davidklaftenegger) [reply src/backend/mpi/write_buffer.hpp]: I think we can leave it to the compiler to optimize away the double not-empty check.

File: src/backend/mpi/write_buffer.hpp
Comment (davidklaftenegger) [reply src/backend/mpi/write_buffer.hpp:174]: This is essentially the same issue as overhauling the statistics then. Fine, let's keep it like this for now.

**Review by davidklaftenegger — DISMISSED** [review]
The patch achieves what it sets out to do, and there are no obvious problems.

**Review by davidklaftenegger — APPROVED** [review]
only indentation changed since previous review.

**Pull request merged**
